### PR TITLE
add libethcore folder

### PR DESCRIPTION
**Pull request by JimmyShi22**

### CODE REVIEW

File: libethcore/BasicAuthority.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
```
Comment (wheatli) [inline libethcore/BasicAuthority.cpp]: I have no idea about the function of this file, mayby we don't need it

File: libethcore/CMakeLists.txt [hunk 2]
```diff
@@ -2,4 +2,5 @@ file(GLOB sources "*.cpp" "*.h")
 
 add_library(ethcore ${sources})
 
-target_link_libraries(ethcore PUBLIC devcore)
+target_include_directories(ethcore PRIVATE ../utils)
```
Comment (wheatli) [inline libethcore/CMakeLists.txt]: no need to add this include, cause it is in the root CMakeList.txt

File: libethcore/ChainOperationParams.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    maxGasLimit("0x7fffffffffffffff"),
+    gasLimitBoundDivisor(0x0400),
+    networkID(0x0),
+    minimumDifficulty(0x020000),
+    difficultyBoundDivisor(0x0800),
+    durationLimit(0x0d)
+{}
+
+EVMSchedule const& ChainOperationParams::scheduleForBlockNumber(u256 const& _blockNumber) const
+{
```
Comment (wheatli) [inline libethcore/ChainOperationParams.cpp]: we don't need to the block number check logic , pls define a BCOSSchedule in EVMSchedule and set it as latest version, such as ByzantiumSchedule

File: libethcore/ChainOperationParams.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        return EIP158Schedule;
+    else if (_blockNumber >= EIP150ForkBlock)
+        return EIP150Schedule;
+    else if (_blockNumber >= homesteadForkBlock)
+        return HomesteadSchedule;
+    else
+        return FrontierSchedule;
+}
+
+u256 ChainOperationParams::blockReward(EVMSchedule const& _schedule) const
```
Comment (wheatli) [inline libethcore/ChainOperationParams.cpp]: delete it

File: libethcore/ChainOperationParams.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+{
+    if (_schedule.blockRewardOverwrite)
+        return *_schedule.blockRewardOverwrite;
+    else
+        return m_blockReward;
+}
+
+void ChainOperationParams::setBlockReward(u256 const& _newBlockReward)
```
Comment (wheatli) [inline libethcore/ChainOperationParams.cpp]: delete it

File: libethcore/ChainOperationParams.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+ * @author Gav Wood <i@gavwood.com>
+ * @date 2015
+ */
+
+#pragma once
+
+#include <libdevcore/Common.h>
+#include <libethcore/Precompiled.h>
+
+#include "Common.h"
```
Comment (wheatli) [inline libethcore/ChainOperationParams.h]: adjust your including sequence

File: libethcore/ChainOperationParams.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    EVMSchedule const& scheduleForBlockNumber(u256 const& _blockNumber) const;
+    u256 blockReward(EVMSchedule const& _schedule) const;
+    void setBlockReward(u256 const& _newBlockReward);
+    u256 maximumExtraDataSize = 1024;
+    u256 accountStartNonce = 0;
+    bool tieBreakingGas = true;
+    u256 minGasLimit;
+    u256 maxGasLimit;
+    u256 gasLimitBoundDivisor;
+    u256 homesteadForkBlock = c_infiniteBlockNumer;
```
Comment (wheatli) [inline libethcore/ChainOperationParams.h]: pls check every member, add some comments for them, delete if useless

File: libethcore/Common.cpp [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file Common.cpp
+ * @author Gav Wood <i@gavwood.com>
+ * @date 2014
+ */
+
+#include "Common.h"
+#include <libdevcore/Base64.h>
+//#include <libdevcore/Terminal.h>
```
Comment (wheatli) [inline libethcore/Common.cpp]: drop it

File: libethcore/Common.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+#include "BlockHeader.h"
+#include "Exceptions.h"
+#include <libdevcore/CommonIO.h>
+#include <libdevcore/SHA3.h>
+#include <libdevcore/easylog.h>
```
Comment (wheatli) [inline libethcore/Common.cpp]: use format tools, make sure your code is standard

File: libethcore/Common.cpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+
+namespace dev
+{
+namespace eth
+{
+const unsigned c_protocolVersion = 63;
+#if ETH_FATDB
```
Comment (wheatli) [inline libethcore/Common.cpp]: delete useless macro

File: libethcore/Common.cpp [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+#else
+const unsigned c_minorProtocolVersion = 2;
+const unsigned c_databaseBaseVersion = 9;
+const unsigned c_databaseVersionModifier = 0;
+#endif
+
+const unsigned c_databaseVersion =
+    c_databaseBaseVersion + (c_databaseVersionModifier << 8) + (23 << 9);
+
+const Address c_blockhashContractAddress(0xf0);
```
Comment (wheatli) [inline libethcore/Common.cpp]: add some comments of this c_blockhashContractAddress

File: libethcore/Common.cpp [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+void badBlock(bytesConstRef _block, string const& _err)
+{
+    BlockHeader bi;
+    DEV_IGNORE_EXCEPTIONS(bi = BlockHeader(_block));
+    badBlockInfo(bi, _err);
+}
+
+string TransactionSkeleton::userReadable(bool _toProxy,
```
Comment (wheatli) [inline libethcore/Common.cpp]: why do you place this method here ?

File: libethcore/ICAP.h [hunk 13]
```diff
@@ -0,0 +1,135 @@
+/*
```
Comment (wheatli) [inline libethcore/ICAP.h]: drop it

File: libethcore/KeyManager.h [hunk 14]
```diff
@@ -0,0 +1,227 @@
+/*
```
Comment (wheatli) [inline libethcore/KeyManager.h]: drop

File: libethcore/Precompiled.cpp [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    bigint const baseLength(parseBigEndianRightPadded(_in, 0, 32));
+    bigint const expLength(parseBigEndianRightPadded(_in, 32, 32));
+    bigint const modLength(parseBigEndianRightPadded(_in, 64, 32));
+
+    bigint const maxLength(max(modLength, baseLength));
+    bigint const adjustedExpLength(expLengthAdjust(baseLength + 96, expLength, _in));
+
+    return multComplexity(maxLength) * max<bigint>(adjustedExpLength, 1) / 20;
+}
+/*
```
Comment (wheatli) [inline libethcore/Precompiled.cpp]: delete them

File: libethcore/SealEngine.cpp [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+        if (_bi.number() && _bi.extraData().size() > chainParams().maximumExtraDataSize)
+        {
+            BOOST_THROW_EXCEPTION(ExtraDataTooBig()
+                                  << RequirementError(bigint(chainParams().maximumExtraDataSize),
+                                         bigint(_bi.extraData().size()))
+                                  << errinfo_extraData(_bi.extraData()));
+        }
+
+        u256 const& daoHardfork = chainParams().daoHardforkBlock;
```
Comment (wheatli) [inline libethcore/SealEngine.cpp]: logic of forks, delete them

File: libethcore/SealEngine.cpp [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        ret->setChainParams(_params);
+    return ret;
+}
+
+EVMSchedule const& SealEngineBase::evmSchedule(u256 const& _blockNumber) const
+{
+    return chainParams().scheduleForBlockNumber(_blockNumber);
+}
+
+u256 SealEngineBase::blockReward(u256 const& _blockNumber) const
```
Comment (wheatli) [inline libethcore/SealEngine.cpp]: drop

File: libethcore/TransactionBase.cpp [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    if (!m_vrs)
+        BOOST_THROW_EXCEPTION(TransactionIsUnsigned());
+
+    if (m_vrs->s > c_secp256k1n / 2)
+        BOOST_THROW_EXCEPTION(InvalidSignature());
+}
+
+void TransactionBase::checkChainId(int chainId) const
+{
+    if (m_chainId != chainId && m_chainId != -4)
```
Comment (wheatli) [inline libethcore/TransactionBase.cpp]: pls get clear about the utility of chainID

File: libethcore/ChainOperationParams.cpp (hunk 3 above)
Comment (wheatli) [reply libethcore/ChainOperationParams.cpp]: and just return BCOSSchedule here what ever block number it is

File: libethcore/Precompiled.cpp (hunk 15 above)
Comment (JimmyShi22) [reply libethcore/Precompiled.cpp]: Has added to our develop document!

File: libethcore/SealEngine.cpp (hunk 16 above)
Comment (JimmyShi22) [reply libethcore/SealEngine.cpp]: delete

File: libethcore/SealEngine.cpp (hunk 17 above)
Comment (JimmyShi22) [reply libethcore/SealEngine.cpp]: deleted

File: libethcore/TransactionBase.cpp (hunk 18 above)
Comment (JimmyShi22) [reply libethcore/TransactionBase.cpp]: Has added to our develop document!

**Pull request merged**